### PR TITLE
Voice Cloning, SSML, Unity UPM, Wyoming Docker 他全実装

### PR DESCRIPTION
**Pull request by ayutaz**
## Summary

6ランタイム横断で Voice Cloning・SSML・Unity UPM など主要機能を一括実装した PR。

### 新機能

- **Voice Cloning (Speaker Encoder)** — ECAPA-TDNN で参照音声から 256 次元 embedding を抽出し、未知話者の声質で合成。Python / Rust / C# / Go / WASM の 5 ランタイム対応
- **SSML 基本サポート** — `<speak>`, `<break>`, `<prosody rate>` を Python / Rust / C# / Go の 4 ランタイムで実装
- **Unity UPM パッケージ** — libpiper_plus P/Invoke + AudioClip API。サンプルシーン 5 種付属
- **MOS ベンチマークツール** — サンプル生成・PESQ/STOI メトリクス計算・調査フォーム生成
- **Wyoming Docker + Home Assistant 統合** — Docker Compose で Wyoming TTS サーバーを起動、HA から利用可能
- **iOS / Android ビルド CI** — libpiper_plus のモバイルクロスコンパイル自動化
- **モデル投稿ガイド** — コミュニティモデル投稿の Issue テンプレート

### 改善・修正

- C API: `PiperPlusSynthOptions` に speaker_embedding フィールド追加（ABI サイズ維持）
- ONNX export: speaker_embedding の embedding dim を固定 256 に明示
- CI gate: `ci-required` で cancelled も失敗扱いに強化
- SSML パーサー: 100KB サイズ制限追加（XML DoS 対策）
- テスト修正 + CI カバレッジ GAP 修正
- 冗長なレガシーテスト 12 件削除（G2P テストスイートに統合済み）
- 完了済みチケット・計画文書の整理削除

### 除去

- **VITS2 Duration Discriminator** — PR #240 で全構成の品質検証を実施し、VITS1 を超えられないことを確認済みのため除去

## 影響範囲

**学習ループへの実質的なロジック変更なし。** Voice Cloning 関連の変更は推論パス (`infer()`, `export_onnx`) のみ。

### Python

| 対象 | ファイル数 | 主な変更 |
|------|----------|---------|
| 推論・エクスポート (`piper_train`) | 10 | Speaker Encoder 新規 6 件, export_onnx / infer_onnx に speaker_embedding 対応, models.py / lightning.py は推論パスとフォーマット変更のみ |
| G2P (`g2p`) | 4 | SSML パーサー新規, テスト, uv.lock |
| Wyoming (`piper_wyoming`) | 3 | Wyoming Protocol TTS サーバー新規 |
| テスト (`tests`) | 9 | Speaker embedding, API, Wyoming handler テスト |
| ランタイム (`python_run`) | 7 | SSML, phonemize, voice, webui lazy import, テスト修正 |

### 他言語ランタイム

| ランタイム | ファイル数 | 主な変更 |
|-----------|----------|---------|
| Rust | 12 | speaker_encoder 新規, engine / voice に speaker_embedding, SSML, テスト |
| C# | 9 | PiperSession / SessionFactory speaker_embedding, SpeakerEncoder 新規, SSML, テスト |
| Go | 9 | engine speaker_embedding, speaker_encoder 新規, SSML, phonemize golden tests |
| C / C++ | 3 | piper_plus.h ABI 維持, C API speaker encoder (experimental) |
| WASM / JS | 4 | speaker-encoder 新規, index.js, 型定義, README |
| Unity | 20 | UPM パッケージ新規 (Runtime 7, Editor 3, Samples 5, Plugins, package.json 等) |

### CI / インフラ

| 対象 | ファイル数 | 主な変更 |
|------|----------|---------|
| GitHub Actions | 17 | ci.yml gate 改善, python-tests / test-webassembly パス追加, release-shared-lib iOS/Android 追加 |
| Issue / PR テンプレート | 5 | model-request, model-submission, config, PR template |
| Docker | 10 | Wyoming 5 件, ollama-stack 4 件, README |
| pytest.ini | 1 | testpaths に `test/`, `tools/benchmark` 追加 |

### ドキュメント / ツール

| 対象 | ファイル数 | 主な変更 |
|------|----------|---------|
| docs/ | 10 | Unity 統合, Wyoming / HA ガイド, benchmark-mos, ORT バージョン仕様, エコシステム調査 |
| tools/benchmark/ | 11 | MOS サンプル生成, メトリクス計算, 調査フォーム, テキスト 6 言語分 |
| test/ | 16 | speaker_encoder fixture / テスト 4 件追加, レガシーテスト 12 件削除 |
| ルート | 3 | CLAUDE.md 更新, CONTRIBUTING_MODELS.md 新規, pytest.ini 更新 |

## Stats

- 178 files changed, +27,267 / -4,202 lines
- 31 commits

## Test plan

- [x] Python テスト — 769 passed
- [x] G2P テスト — 729 passed
- [x] Rust ビルド — OK (warning 0)
- [x] C# テスト — 1,088 passed
- [x] Go テスト — all passed
- [ ] Wyoming Docker ビルド — Docker 未起動のためスキップ

### CODE REVIEW

File: src/python/piper_train/vits/models.py [hunk 1]
```diff
@@ -971,6 +1083,9 @@ def forward(
             l_length = torch.sum((logw - logw_) ** 2, [1, 2]) / torch.sum(
                 x_mask
             )  # for averaging
+            # For VITS2: provide (x_dp_detached, x_mask, logw_real, logw_gen)
+            if self.vits2 and self.dur_disc is not None:
+                dur_info = (x_dp.detach(), x_mask, logw_, logw)
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/vits/models.py]: Fixed in 29cf35c. SDP パスでも dur_info を計算するようにしました。

File: src/python/piper_train/vits/lightning.py [hunk 2]
```diff
@@ -572,6 +598,24 @@ def training_step_g(self, batch: Batch):
...
+                dur_disc = self.model_g.dur_disc
+                # Generator wants dur_disc to classify fake as real
+                # Run dur_disc in forward-only mode for the fake branch
+                disc_fake = dur_disc._forward_single(
+                    x_dp_det, x_mask_dp, logw_g
+                )
+                # BCE loss: generator wants disc_fake -> 1 (real)
+                loss_dur_gen = F.binary_cross_entropy_with_logits(
+                    disc_fake, torch.ones_like(disc_fake)
+                )
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/vits/lightning.py]: Fixed in 29cf35c. Generator step で dur_disc.requires_grad_(False) → forward → requires_grad_(True) の標準 GAN パターンを適用しました。

File: src/csharp/PiperPlus.Core/Inference/PiperSession.cs [hunk 3]
```diff
@@ -259,6 +267,35 @@ public SynthesisResult SynthesizeWithDurations(SynthesisInput input)
...
+                inputValues.Add(speakerEmbMaskTensor);
+            }
+            else
+            {
+                // Model supports it but no embedding provided — send mask=0
+                long[] mask = [0];
+                speakerEmbMaskTensor = OrtValue.CreateTensorValueFromMemory(mask, [1]);
+                inputNames.Add("speaker_embedding_mask");
+                inputValues.Add(speakerEmbMaskTensor);
+            }
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/csharp/PiperPlus.Core/Inference/PiperSession.cs:312]: Fixed in 29cf35c. embedding 未指定時もゼロテンソル (embDim はメタデータから取得、デフォルト256) を提供し、mask を 2D [1,1] に修正しました。

File: src/csharp/PiperPlus.Core/Inference/SessionFactory.cs [hunk 4]
```diff
@@ -272,6 +272,16 @@ public static void Warmup(
...
 
+            // speaker_embedding_mask: int64 [1] = 0 (no embedding during warmup)
+            OrtValue? speakerEmbMaskTensor = null;
+            if (metadata.ContainsKey("speaker_embedding_mask"))
+            {
+                long[] mask = [0];
+                speakerEmbMaskTensor = OrtValue.CreateTensorValueFromMemory(mask, [1]);
+                inputNames.Add("speaker_embedding_mask");
+                inputValues.Add(speakerEmbMaskTensor);
+            }
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/csharp/PiperPlus.Core/Inference/SessionFactory.cs]: Fixed in 29cf35c. Warmup で speaker_embedding + speaker_embedding_mask 両方を提供。mask 形状を 2D [1,1] に修正。dispose も追加しました。

File: src/go/piperplus/engine.go [hunk 5]
```diff
@@ -212,6 +219,41 @@ func (e *OnnxEngine) Synthesize(ctx context.Context, req *SynthesisRequest) (*Sy
...
+			if err != nil {
+				return nil, &InferenceError{Msg: "failed to create speaker_embedding tensor", Err: err}
+			}
+			tensors = append(tensors, embTensor)
+			inputs = append(inputs, embTensor)
+
+			maskTensor, err := ort.NewTensor(ort.NewShape(1), []int64{1})
+			if err != nil {
+				return nil, &InferenceError{Msg: "failed to create speaker_embedding_mask tensor", Err: err}
+			}
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/go/piperplus/engine.go]: Fixed in 29cf35c. mask 形状を ort.NewShape(1) → ort.NewShape(1, 1) に修正し、ONNX export の 2D 定義と一致させました。

File: src/go/piperplus/engine.go [hunk 6]
```diff
@@ -212,6 +219,41 @@ func (e *OnnxEngine) Synthesize(ctx context.Context, req *SynthesisRequest) (*Sy
...
+			if err != nil {
+				return nil, &InferenceError{Msg: "failed to create speaker_embedding placeholder", Err: err}
+			}
+			tensors = append(tensors, placeholderTensor)
+			inputs = append(inputs, placeholderTensor)
+
+			maskTensor, err := ort.NewTensor(ort.NewShape(1), []int64{0})
+			if err != nil {
+				return nil, &InferenceError{Msg: "failed to create speaker_embedding_mask tensor", Err: err}
+			}
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/go/piperplus/engine.go]: Fixed in 29cf35c. 同上、placeholder パスの mask 形状も (1,1) に修正済みです。

File: src/python/piper_train/export_onnx.py [hunk 7]
```diff
@@ -390,6 +417,26 @@ def infer_forward(
...
+    dummy_input_list.append(dummy_speaker_embedding)
+    input_names.append("speaker_embedding")
+    dynamic_axes["speaker_embedding"] = {0: "batch_size", 1: "emb_dim"}
+
+    dummy_input_list.append(dummy_speaker_embedding_mask)
+    input_names.append("speaker_embedding_mask")
+    dynamic_axes["speaker_embedding_mask"] = {0: "batch_size"}
+
+    _LOGGER.info(
+        "Exporting model with speaker_embedding support (emb_dim=%d)",
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/export_onnx.py:438]: Fixed in 29cf35c. axis 1 の dynamic 宣言を削除し、`{0: "batch_size"}` のみに変更。spk_proj の重みは 256 固定のためコメントで明記しました。

[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/export_onnx.py:438]: Fixed in 29cf35c. 上記と同一修正です。

File: com.piper-plus.tts/Runtime/AudioClipExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                Debug.LogWarning("[PiperPlus] Cannot create AudioClip from empty samples.");
+                return null;
+            }
+
+            const float scale = 1.0f / 32767.0f;
+            float[] floatSamples = new float[samples.Length];
+            for (int i = 0; i < samples.Length; i++)
+            {
+                floatSamples[i] = samples[i] * scale;
+            }
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply com.piper-plus.tts/Runtime/AudioClipExtensions.cs]: Fixed in 29cf35c. 除数を 32767 → 32768f に変更し、-32768 が正確に -1.0f にマッピングされるようにしました。

File: com.piper-plus.tts/Editor/PiperPlusEditor.cs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+                    // Schedule cleanup after clip finishes
+                    float duration = clip.length + 0.5f;
+                    EditorApplication.delayCall += () =>
+                    {
+                        EditorApplication.delayCall += () =>
+                        {
+                            if (go != null)
+                                Object.DestroyImmediate(go);
+                        };
+                    };
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply com.piper-plus.tts/Editor/PiperPlusEditor.cs]: Fixed in 29cf35c. delayCall を EditorApplication.update コールバックに変更し、timeSinceStartup + isPlaying で再生完了まで待機してから GameObject を破棄するようにしました。

File: src/cpp/piper_plus.h [hunk 10]
```diff
@@ -93,13 +93,15 @@ typedef struct PiperPlusConfig {
...
-    int32_t _reserved[8];         /* Must be zero */
+    int32_t speaker_id;                 /* Speaker index (default: 0) */
+    int32_t language_id;                /* Language index (-1 = auto-detect, default: -1) */
+    float   noise_scale;                /* VITS noise_scale (default: 0.667) */
+    float   length_scale;               /* VITS length_scale (default: 1.0) */
+    float   noise_w;                    /* VITS noise_w (default: 0.8) */
+    float   sentence_silence_sec;       /* Silence between sentences in sec (default: 0.2) */
+    const float *speaker_embedding;     /* Voice cloning: float32 embedding (NULL = use speaker_id) */
+    int32_t      speaker_embedding_dim; /* Number of elements in speaker_embedding (0 = disabled) */
+    int32_t _reserved[6];               /* Must be zero */
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/cpp/piper_plus.h]: Fixed in 29cf35c. _reserved[6] → _reserved[5] に変更し、構造体の合計サイズを元の 56 bytes に維持しました (pointer 8B + int32 4B = 12B 追加分を _reserved から 3 スロット削減)。

File: src/cpp/piper_plus_c_api.cpp [hunk 11]
```diff
@@ -1063,4 +1072,82 @@ PIPER_PLUS_API const char *piper_plus_available_languages(PiperPlusEngine *engin
     return engine->availableLanguagesStr.c_str();
 }
 
+// ===== M3-04: Speaker Encoder for Voice Cloning =====
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/cpp/piper_plus_c_api.cpp]: Fixed in 29cf35c. ヘッダーと実装に `EXPERIMENTAL -- not yet implemented` を明記し、エラーメッセージも更新しました。API surface は forward compatibility のため維持しています。

File: src/cpp/piper_plus_c_api.cpp [hunk 12]
```diff
@@ -1063,4 +1072,82 @@ PIPER_PLUS_API const char *piper_plus_available_languages(PiperPlusEngine *engin
...
+    } catch (const std::exception &e) {
+        set_error(e.what());
+        return nullptr;
+    } catch (...) {
+        set_error("Unknown error creating speaker encoder");
+        return nullptr;
+    }
+}
+
+PIPER_PLUS_API int32_t piper_plus_speaker_encoder_encode(
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/cpp/piper_plus_c_api.cpp:1111]: Fixed in 29cf35c. 上記と同一修正です。

File: src/cpp/piper_plus_c_api.cpp [hunk 13]
```diff
@@ -1063,4 +1072,82 @@ PIPER_PLUS_API const char *piper_plus_available_languages(PiperPlusEngine *engin
...
+    }
+    if (sample_rate <= 0) {
+        set_error("sample_rate must be positive");
+        return -1;
+    }
+
+    // NOTE: Full inference implementation depends on the SpeakerEncoder class
+    // from M3-01. This validates the API surface and returns a placeholder.
+    set_error("Speaker encoder inference not yet connected to M3-01 backend");
+    return -1;
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/cpp/piper_plus_c_api.cpp]: Fixed in 29cf35c. 上記と同一修正です。

File: src/python/g2p/piper_plus_g2p/ssml.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+        list[SSMLSegment]
+            Ordered segments ready for phonemization.
+        """
+        if not SSMLParser.is_ssml(ssml_text):
+            # Plain text -- return as a single segment.
+            return [SSMLSegment(text=ssml_text)]
+
+        try:
+            root = ET.fromstring(ssml_text)  # noqa: S314
+        except ET.ParseError:
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/piper_plus_g2p/ssml.py:124]: Fixed in 29cf35c. _MAX_SSML_SIZE = 100_000 (100KB) のサイズ制限を parse() の冒頭に追加しました。defusedxml は依存に含まれていないため、新規依存追加なしでの対策としています。テスト3件追加済み。

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -339,3 +418,31 @@ jobs:
...
+      - name: Check all job results
+        run: |
+          results=("${{ needs.cpp-tests.result }}" "${{ needs.csharp-tests.result }}" "${{ needs.csharp-build.result }}" "${{ needs.python-tests.result }}" "${{ needs.rust-tests.result }}" "${{ needs.npm-package-tests.result }}" "${{ needs.lint.result }}")
+          for result in "${results[@]}"; do
+            if [[ "$result" == "failure" ]]; then
+              echo "One or more jobs failed"
+              exit 1
+            fi
+          done
+          echo "All jobs passed or were skipped"
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply .github/workflows/ci.yml:448]: Fixed in 29cf35c. 条件を == 'failure' から != 'success' && != 'skipped' に変更し、cancelled を含む全非成功ステータスで gate が失敗するようにしました。
